---
string software_version
string hardware_version
string[] extension